firebase-firestore/src/main/java/com/google/firebase/firestore/model/ObjectValue.java

```diff
@@ -103,32 +109,41 @@ private FieldMask extractFieldMask(MapValue value) {
   }
 
   @Nullable
-  private Value extractNestedValue(Value value, FieldPath fieldPath) {
+  private static Value extractNestedValue(Value value, FieldPath fieldPath) {
     if (fieldPath.isEmpty()) {
       return value;
     } else {
       for (int i = 0; i < fieldPath.length() - 1; ++i) {
-        value = value.getMapValue().getFieldsOrDefault(fieldPath.getSegment(i), null);
-        if (!Values.isMapValue(value)) {
+        value = extractSegmentValue(value, fieldPath.getSegment(i));
+        if (value == null) {
           return null;
         }
       }
-      return value.getMapValue().getFieldsOrDefault(fieldPath.getLastSegment(), null);
+      return extractSegmentValue(value, fieldPath.getLastSegment());
     }
   }
 
+  private static Value extractSegmentValue(Value value, String segment) {
+    return Values.isMapValue(value) ? value.getMapValue().getFieldsOrDefault(segment, null) : null;
```

Comment:
Was it intentional to invert the logic for extracting a value from a MapValue? Previously the code ran `Values.isMapValue(value)` on the `value` contained in the input map and returned `null` if it was not a map. Now that check is not performed on the extracted value. Seems like this could change what this method returns?

Reply:
Semantically, this should be equivalent. The helper function is used in other contexts where value might be `null`, so I had to add check. Since this is being done in helper, it seemed like overkill to also do it in the loop. Technically the loop will work fine even without early return, but the short circuiting behavior has nice performance characteristics.

It isn't the easiest code to read, so any recommendations are welcome.

Reply:
I think the new code already is more readable. So thanks for the improvement, and clarifying comment. I don't have any recommendations for improvement, but consider a unit test if you have any concerns about equivalency.